pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,667 @@
+						continue
+					}
+					switch dst.t {
+					case DestinationS3:
+						fmt.Fprintf(o.Out, "%s s3://%s:%s\n", toDigest, dst.ref, tag)
+					default:
+						fmt.Fprintf(o.Out, "%s %s:%s\n", toDigest, dst.ref, tag)
+					}
+				}
+				if len(dst.tags) == 0 {
+					toDigest, err := toManifests.Put(ctx, srcManifest)
+					if err != nil {
+						digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to push manifest to %s: %v", dst.ref, err)})
+						continue
+					}
+					switch dst.t {
+					case DestinationS3:
+						fmt.Fprintf(o.Out, "%s s3://%s\n", toDigest, dst.ref)
+					default:
+						fmt.Fprintf(o.Out, "%s %s\n", toDigest, dst.ref)
+					}
+				}
+			}
+		}
+		for _, err := range append(tagErrs, digestErrs...) {
+			errs = append(errs, err)
+		}
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func processManifestList(ctx apirequest.Context, srcDigest digest.Digest, srcManifest distribution.Manifest, manifests distribution.ManifestService, ref imageapi.DockerImageReference, filterFn func(*manifestlist.ManifestDescriptor) bool) ([]distribution.Manifest, digest.Digest, error) {
+	var srcManifests []distribution.Manifest
+	switch t := srcManifest.(type) {
+	case *manifestlist.DeserializedManifestList:
+
+		filtered := make([]manifestlist.ManifestDescriptor, 0, len(t.Manifests))
+		for _, manifest := range t.Manifests {
+			if !filterFn(&manifest) {
+				glog.V(5).Infof("Skipping image for %#v from %s", manifest.Platform, ref)
+				continue
+			}
+			glog.V(5).Infof("Including image for %#v from %s", manifest.Platform, ref)
+			filtered = append(filtered, manifest)
+		}
+
+		if len(filtered) == 0 {
+			return nil, "", nil
+		}
+
+		// if we're filtering the manifest list, update the source manifest and digest
+		if len(filtered) != len(t.Manifests) {
+			t, err := manifestlist.FromDescriptors(filtered)
+			if err != nil {
+				return nil, "", fmt.Errorf("unable to filter source image %s manifest list: %v", ref, err)
+			}
+			_, body, err := t.Payload()
+			if err != nil {
+				return nil, "", fmt.Errorf("unable to filter source image %s manifest list (bad payload): %v", ref, err)
+			}
+			srcManifest = t
+			srcDigest = srcDigest.Algorithm().FromBytes(body)
+		}
+
+		for i, manifest := range t.Manifests {
+			childManifest, err := manifests.Get(ctx, manifest.Digest, schema2ManifestOnly)
+			if err != nil {
+				return nil, "", fmt.Errorf("unable to retrieve source image %s manifest #%d from manifest list: %v", ref, i+1, err)
+			}
+			srcManifests = append(srcManifests, childManifest)
+		}
+
+		switch {
+		case len(srcManifests) == 1:
+			_, body, err := srcManifests[0].Payload()
+			if err != nil {
+				return nil, "", fmt.Errorf("unable to convert source image %s manifest list to single manifest: %v", ref, err)
+			}
+			return srcManifests, srcDigest.Algorithm().FromBytes(body), nil
+		default:
+			return append(srcManifests, srcManifest), srcDigest, nil
```

Comment:
The manifest itself needs to be modified to omit all the references filtered out. Otherwise the manifest Put will fail due to not existing references in the destination repository.

Reply:
The manifest is modified.

Reply:
The manifest list is modified.

Reply:
Sorry, I was confused by the name `srcManifest` which is the input parameter. Having different name for the processed manifest list wouldn't hurt.